tensorflow_probability/python/stats/sample_stats.py

```diff
@@ -645,6 +807,104 @@ def variance(x, sample_axis=0, keepdims=False, name=None):
         x, y=None, sample_axis=sample_axis, event_axis=None, keepdims=keepdims)
 
 
+VarianceState = collections.namedtuple('VarianceState',
+                                       'num_samples, mean, m2, dtype')
+
+
+def init_running_variance(shape=(), dtype=tf.float32):
+  """Initialize the Running Variance State.
+
+  Args:
+    shape: Python `Tuple` or `TensorShape` representing the shape of
+      incoming samples. Scalar samples are assumed by default.
+    dtype: Dtype of incoming samples and the resulting statistics.
+      By default, the dtype is `tf.float32`
+
+  Returns:
+    state: `VarianceState` representing a stream of no inputs
+  """
+  return VarianceState(
+      num_samples=tf.zeros((), dtype=dtype),
+      mean=tf.zeros(shape, dtype=dtype),
+      m2=tf.zeros(shape, dtype=dtype),
+      dtype=dtype,
+  )
+
+
+def update_running_variance(curr_state, new_sample, axis=None):
+  """Update the streaming statistics of the `VarianceState` with a new sample
+  using Welford's Online Algorithm.
+
+  Args:
+    curr_state: `VarianceState` that depicts the current state of streaming
+      statistics
+    new_sample: Incoming sample with shape and dtype compatible with those used
+      to form the `VarianceState` by `init_running_variance`
+    axis: If chunking is desired, this is an Integer that specifies the axis
+      with chunked samples. For individual samples, this is None. By default,
+      samples are not chunked (`axis` is None)
+
+  Returns:
+    state: `VarianceState` with updated calculations
+  """
+  if tf.is_tensor(new_sample):
+    new_sample = tf.cast(new_sample, dtype=curr_state.dtype)
+  else:
+    new_sample = tf.convert_to_tensor(new_sample, dtype=curr_state.dtype)
+  if axis is not None:
+    b_n = tf.cast(prefer_static.shape(new_sample)[axis], dtype=curr_state.dtype)
+    b_mean = tf.math.reduce_mean(new_sample, axis=axis)
+    b_m2 = tf.reduce_sum(
+        tf.math.squared_difference(
+            tf.expand_dims(
+                b_mean,
+                axis=axis),
+            new_sample),
+        axis=axis)
+  else:
+    b_n = tf.ones((), dtype=curr_state.dtype)
+    b_mean = new_sample
+    b_m2 = tf.zeros(prefer_static.shape(curr_state.m2), dtype=curr_state.dtype)
+
+  total_n = curr_state.num_samples + b_n
+  delta = b_mean - curr_state.mean
+  new_mean = curr_state.mean + delta * (b_n / total_n)
+  new_m2 = (curr_state.m2 + b_m2
```

Comment:
You might actually note in the docstring that this works for batches or individual updates. I guess instead of the typical:

```python
old_delta = (x[t] - mean[t])
mean[t+1] = mean[t] + old_delta / t
new_delta = (x[t] - mean[t + 1])
raw_var[t + 1] = raw_var[t] + old_delta * new_delta
```

you are using

```python
delta = (x[t] - mean[t])
mean[t+1] = mean[t] + delta / t
raw_var[t + 1] = raw_var[t] + delta * delta * t / t + 1
```
and `b_m2` is a correction for batches?

Reply:
Yup, pretty much. Though instead of calculating the raw variance, I just calculate the sum of squared distances from the mean (i.e. no  t / t + 1 term which is handled in finalize); hence, m2 is meant to represent that